core/src/main/java/io/questdb/griffin/engine/orderby/LimitedSizePartiallySortedLightRecordCursor.java

```diff
@@ -0,0 +1,207 @@
+            long skipFirst,
+            long skipLast,
+            int timestampIndex
+    ) {
+        this.chain = chain;
+        this.comparator = comparator;
+        this.chainCursor = chain.getCursor();
+        this.limit = limit;
+        this.skipFirst = skipFirst;
+        this.skipLast = skipLast;
+        this.isOpen = true;
+        this.timestampIndex = timestampIndex;
+    }
+
+    @Override
+    public void close() {
+        if (isOpen) {
+            Misc.free(chain);
+            Misc.free(base);
+            isOpen = false;
+        }
+    }
+
+    @Override
+    public Record getRecord() {
+        return baseRecord;
+    }
+
+    @Override
+    public Record getRecordB() {
+        return base.getRecordB();
+    }
+
+    @Override
+    public SymbolTable getSymbolTable(int columnIndex) {
+        return base.getSymbolTable(columnIndex);
+    }
+
+    @Override
+    public boolean hasNext() {
+        if (!isChainBuilt) {
+            buildChain();
+            isChainBuilt = true;
+        }
+        if (rowsLeft-- > 0 && chainCursor.hasNext()) {
+            base.recordAt(baseRecord, chainCursor.next());
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public SymbolTable newSymbolTable(int columnIndex) {
+        return base.newSymbolTable(columnIndex);
+    }
+
+    @Override
+    public void of(RecordCursor base, SqlExecutionContext executionContext) {
+        if (!isOpen) {
+            chain.reopen();
+            isOpen = true;
+        }
+
+        this.base = base;
+        baseRecord = base.getRecord();
+        circuitBreaker = executionContext.getCircuitBreaker();
+        isChainBuilt = false;
+        rowsInGroup = 0;
+        rowsSoFar = 0;
+        timestampValue = Numbers.LONG_NaN;
+        timestampInitialized = false;
+        chain.clear();
+    }
+
+    @Override
+    public void recordAt(Record record, long atRowId) {
+        base.recordAt(record, atRowId);
+    }
+
+    @Override
+    public long size() {
+        return isChainBuilt ? Math.max(chain.size() - skipFirst - skipLast, 0) : -1;
+    }
+
+    @Override
+    public void toTop() {
+        chainCursor.toTop();
+
+        long skipLeft = skipFirst;
+        while (skipLeft-- > 0 && chainCursor.hasNext()) {
+            chainCursor.next();
+        }
+
+        rowsLeft = Math.max(chain.size() - skipFirst - skipLast, 0);
+    }
+
+    private void buildChain() {
+        final Record placeHolderRecord = base.getRecordB();
+        if (limit != 0) {
+            // first record ever, we've to find the timestamp value
+            if (!timestampInitialized) {
+                if (base.hasNext()) {
+                    chain.put(
+                            baseRecord,
+                            base,
+                            placeHolderRecord,
+                            comparator
+                    );
+                    timestampValue = baseRecord.getTimestamp(timestampIndex);
```

Comment:
nit: considering `rowsInGroup` field, it might be more obvious to rename `timestampValue` to `groupTimestamp` or something like this.

Reply:
Changed in latest commit